### PR TITLE
Feature/uppsf 2667 dynamic relations

### PR DESCRIPTION
**Pull request by ivan-p-nikolov**
# Description

## What

Add functionality for dynamic relationships

## Why

https://financialtimes.atlassian.net/browse/UPPSF-2667

## Anything, in particular, you'd like to highlight to reviewers

Mention here sections of code which you would like reviewers to pay extra attention to .E.g

_Would appreciate a second pair of eyes on the test_  
_I am not quite sure how this bit works_  
_Is there a better library for doing x_  

## Scope and particulars of this PR (Please tick all that apply)

- [ ] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [ ] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: ontology/aggregate.go [hunk 1]
```diff
@@ -94,12 +94,22 @@ func mergeCanonicalInformation(c ConcordedConcept, s SourceConcept, scopeNoteOpt
...
 	c.Aliases = append(c.Aliases, s.PrefLabel)
-	for key := range GetConfig().FieldToNeoProps {
-		val, has := s.Fields[key]
-		if !has {
-			continue
+
+	for key, val := range s.Fields {
+		if GetConfig().HasField(key) {
+			c.Fields[key] = val
+		}
```
Comment (ivanruski) [inline ontology/aggregate.go:101]: **suggestion(non-blocking):** We can place `continue` here to show intent that if we encounter valid field, we don't expect anything else to happen.

**Review by ivanruski — APPROVED** [review]
LGTM. Just a non-blocking **suggestion** and a **thought**

File: ontology/config.go [hunk 2]
```diff
@@ -42,4 +139,15 @@ func GetConfig() Config {
 
 func setGlobalConfig(cfg Config) {
 	config = cfg
+	initConfig(&config)
+}
+
+func initConfig(cfg *Config) {
+	cfg.MergingStrategies = map[string]MergingStrategy{}
+	for _, rel := range cfg.Relationships {
+		if rel.Strategy == "" {
```
Comment (epavlova) [inline ontology/config.go:148]: **thought:** Do you think it could be clearer if we have default merging strategy?

**Review by epavlova — COMMENTED** [review]
I'm okay with approving the PR, just still wondering on the merging strategies question.

File: ontology/aggregate.go [hunk 3]
```diff
@@ -94,12 +94,22 @@ func mergeCanonicalInformation(c ConcordedConcept, s SourceConcept, scopeNoteOpt
...
+
+		switch GetConfig().MergingStrategies[key] {
+		case OverrideStrategy:
+			c.Fields[key] = val
```
[the remark this answers is not in the document]
Comment (ivan-p-nikolov) [reply ontology/aggregate.go:105]: The strategy is determined by the old implementation. For instance `SupersededByUUIDs` [overwritten](https://github.com/Financial-Times/aggregate-concept-transformer/blob/0f149f2e21442848bcb60a416bb86bca33d7909e/ontology/aggregate.go#L138). And `MembershipRoles` is [aggregated](https://github.com/Financial-Times/aggregate-concept-transformer/blob/0f149f2e21442848bcb60a416bb86bca33d7909e/ontology/aggregate.go#L194).

In general, I'm not completely sure why we collect relations in the concorded concept. And I don't expect future relations to use any form of `MergingStrategy`.

File: ontology/config.go (hunk 2 above)
Comment (ivan-p-nikolov) [reply ontology/config.go:148]: I'm not sure what is your vision for default strategy. 

Variant A: Where if it is not set, we set it to the default `IgnoreStarategy`.
```golang
for _, rel := range cfg.Relationships {
	strategy := rel.Strategy
	if strategy == "" {
		strategy = IgnoreStarategy
	}
	cfg.MergingStrategies[rel.ConceptField] = strategy
}
```
Or

Variant B: if it is not set, we alert.
```golang
for relName, rel := range cfg.Relationships {
	if rel.Strategy == "" {
		panic(fmt.Errorf("relation %s doesn't have a strategy",relName))
	}
	cfg.MergingStrategies[rel.ConceptField] = rel.Strategy
}
```

Also maybe I should check if the property is "not supported" strategy.

**Review by ivanruski — APPROVED** [review]

**Review by epavlova — APPROVED** [review]

**Pull request merged**